[RFC] Added database support for the binary-uuid type

AntonNguyen opened the pull request:
I haven't worked with binary uuids before, so it'd be great if I could get some extra eyes on this. I was just following the thread in #10432, and decided to take a stab at it. 😄 

refs #10432

AntonNguyen committed:
Added database support for the binary-uuid type
refs #10432

markstory commented on src/Database/Type/BinaryUuidType.php:
```diff
@@ -0,0 +1,165 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.0.0
```
Should be 3.6.0

markstory commented on tests/TestCase/Database/Type/BinaryUuidType.php:
```diff
@@ -0,0 +1,113 @@
+
+        $fh = fopen(__FILE__, 'r');
+        $result = $this->type->toDatabase($fh, $this->driver);
+        $this->assertSame($fh, $result);
```
Should there be a test for a string uuid coming back as bytes?

markstory reviewed (commented):
Thanks for taking this on. A few people should be quite happy🎉

markstory committed:
Rename files and fix since tag.

markstory committed:
Cast string UUID to binary ones when preparing data.

If users pass in string uuids we should automatically convert that data
to its binary representation.

markstory committed:
Add binaryuuid reflection and generation to sqlite driver.

markstory committed:
Add binaryuuid reflection/generation to Mysql schema.

markstory committed:
Add binaryuuid reflection/generation to sqlserver.

markstory committed:
Add type mapping for binary uuids.

markstory committed:
Add integration test for binary uuid.

Extend the existing uuid table test to do the same tests for binary
uuids.

markstory committed:
Add binaryuuid support for postgres.

Postgres UUIDs are _already_ binary compatible. In postgres binaryuuid
and standard uuid map to the same SQL data type. Postgres does not
support lengths on its BYTEA field so UUID is our best option here.

markstory commented:
Updated with schema reflection for each db driver and integrated the new type into the Type mappings. In MySQL, SQLServer and SQLite `BINARY(16)` will now map to `binaryuuid`. Postgres does not support binary columns with lengths so binaryuuid maps to the `UUID` datatype there.

markstory committed:
Fix PHPCS error.

markstory committed:
SQLServer lengths are stringy.

markstory committed:
Map binaryuuid into UNIQUEIDENTIFIER

Like postgres, SQLServer will take binary data in its UNIQUEIDENTIFIER
data and correctly preserve it as a UUID under the hood. This means
there is no reflection arm for binary uuids as they aren't a different
datatype in sqlserver.

markstory committed:
Make test more generalized.

Don't use a hardcoded key to test deletes.

markstory committed:
Add additional test for get() with uuid.

markstory commented:
Thanks @AntonNguyen :clap:

dereuromark commented:
@AntonNguyen Do you agree on the following simpifications and fixes? https://github.com/dereuromark/cakephp-shim/pull/11 We could make a follow up PR for 3.6 then.

dereuromark commented:
Also, when adding invalid UUIDs like `'test'` it will save them as `0e000000000000000000000000000000
` instead of failing, and also a lot of

    Warning (2): pack(): Type H: illegal hex digit t [ROOT/.../BinaryUuidType.php, ...]

etc
But maybe this should be part of the validation/patching part and is out of scope of the type class then?

markstory commented:
> But maybe this should be part of the validation/patching part and is out of scope of the type class then?

Yes that seems like a validation concern to me.

lorenzo commented:
I just saw this... would have preferred a sorted binary uuid implementation, like the one we have in some of the tests.

I guess that could be another custom type :)

dereuromark commented:
Do you mean https://github.com/alizain/ulid or https://github.com/mbbender/doctrine-ordered-uuid-generator/blob/master/src/ORM/Id/OrderedGuidGenerator.php @lorenzo  ?

markstory commented:
@lorenzo We can still change this implementation as it is in the unreleased 3.next branch.

dereuromark commented:
Is there any interest to finish this for 3.x or 4.x?
